### PR TITLE
Add source level debugging guide

### PR DESCRIPTION
**Pull request by donnayoyo**
Add 'Source Level Debugging' page under 'Developer's Guide'

### CODE REVIEW

**Comment by donnayoyo** [comment]
Since I don't have right to add 'reviwer', hopefully, the comments will direct you to review the code:
@mborgerson
@myguitar 
@huang-jin-intel 

Thanks!

File: source/developer-guides/debug.rst [hunk 1]
```diff
@@ -0,0 +1,197 @@
+.. _debug:
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: Rename 'debug' to 'debugging-with-udk'. Same for file name to 'debugging-with-udk.rst'

File: source/developer-guides/debug.rst [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+Source Level Debugging
+--------------
+
+This explains how to perform basic debugging operation with Intel UDK Debugger on SBL over serial port. (used Apollo Lake (ABL) as example)
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: extend underline all the way wider than the header above.

Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: * Use substitution token for 'SBL': |SPN|. There are many examples in the source.
* Remove (...) as the content is fairly generic overall. We could add more concrete steps for UP2 later.

File: source/developer-guides/index.rst [hunk 3]
```diff
@@ -23,5 +23,6 @@ Developer's Guide
    versioning
    boot-performance
    shell
+   debug
```
Comment (huang-jin-intel) [inline source/developer-guides/index.rst]: rename to 'debugging-with-udk'

File: source/developer-guides/debug.rst [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+
+Prerequisite
+^^^^^^^^^^^^^^
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: * The hyperlinks in this section are not used properly. There are many examples in up2.rst. Please use it as reference to update this section.
* WinDBG requires a specific version. State we have tested with version: https://github.com/pyrasis/windowsprojectbook/blob/master/Applications/Debugging%20Tools%20for%20Windows/dbg_x86_6.11.1.404.msi

File: source/developer-guides/debug.rst [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+.. |Windows Debug Tool(WinDbg)| raw:: html
+
+   <a href="https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools" target="_blank">Windows Debug Tool(WinDbg)</a>  
+
+General Steps
+^^^^^^^^^^^^^^
+
+**Step 1 - Enable Debug Agent**
+
+Enable DebugAgent on SBL and flash the SBL image(or IFWI) on TARGET. By default, DebugAgent is not enabled. It can be simply enabled with a single line change.
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: For SBL and APL references, use corresponding name in substitutions.txt

File: source/developer-guides/debug.rst [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    self.ENABLE_SOURCE_DEBUG = 1
+  
+2. APL platform requires to set **SKIP_STAGE1A_SOURCE_DEBUG** to **1**:: 
+
+    self.SKIP_STAGE1A_SOURCE_DEBUG = 1
+    
+    
+.. note::
+    On ABL platform, the Stage1A has size limitation for debugging which require set SKIP_STAGE1A_SOURCE_DEBUG to 1
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: I assume you meant Apollo Lake, not ABL.

File: source/developer-guides/debug.rst [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    
+  For TCP port configuration::
+  
+    [Debug Port]
+    Channel = TCP
+    Server = 127.0.0.1
+    Port = 1234
+
+|
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: Remove extra spacer here

File: source/developer-guides/debug.rst [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+    [Maintenance]
+    Trace=0x10          <== This is optional. 0x0: Disable trace output, 0x3f: Enable full trace output
+
+|
+
+**Step 3 - Launch Intel UDK Debugger on Host**
+
+* Windows
+ - Launch "**Start WinDbg with Intel UDK Debugger Tool**"
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: Launch '...' from Windows Start Menu

File: source/developer-guides/debug.rst [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  * Linux::
+  
+      qemu-system-x86_64 -m 256M -machine q35 -nographic -pflash Outputs/qemu/SlimBootloader.bin -serial tcp:127.0.0.1:1234,server
+
+
+|
+
+**Step 5 - Start debugging**
+
+Start debugging with WinDbg or GDB
```
Comment (huang-jin-intel) [inline source/developer-guides/debug.rst]: Add a tip here: 

Sometimes UDK tool does not terminate itself completely after exit. Look for process name ‘eXDI’ and kill it before starting a new debugging session.

File: source/developer-guides/debug.rst (hunk 1 above)
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

File: source/developer-guides/debug.rst (hunk 2 above)
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

File: source/developer-guides/index.rst (hunk 3 above)
Comment (donnayoyo) [reply source/developer-guides/index.rst]: fixed

File: source/developer-guides/debug.rst [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+    * Microsoft Win7 x64 or Win8 x64  or Win10 x64
+    
+    * Microsoft Windows Debug Tool(WinDbg) for proper Windows version.     
+  
+
+.. |Intel UDK Debugger| raw:: html
+
+   <a href="https://firmware.intel.com/develop/intel-uefi-tools-and-utilities/intel-uefi-development-kit-debugger-tool#overlay-context=develop" target="_blank">Intel UDK Debugger on HOST</a>
+
```
[the remark this answers is not in the document]
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

File: source/developer-guides/debug.rst (hunk 5 above)
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

File: source/developer-guides/debug.rst (hunk 6 above)
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

File: source/developer-guides/debug.rst (hunk 7 above)
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

File: source/developer-guides/debug.rst (hunk 8 above)
Comment (donnayoyo) [reply source/developer-guides/debug.rst]: fixed

**Pull request merged**